File: ruby/maybe_hugs.rb
# Maybe hug a cutie
class MaybeHug
  attr_reader :cutie

  def initialize(cutie)
    @cutie = cutie
  end

  def hug
    if cutie.respond_to? :hug
      cutie.hug
    else
      'Cutie prefers empathy!'
    end
  end
end

File: ocaml/maybehugs.ml
type cuttie = Hug of string | Empathy of string

let empathic = Empathy "Cutie prefers empathy! "
let huggable = Hug "HUG !"

let hug = function
    (Hug(s))     -> s
  | (Empathy(s)) -> s

let katie = huggable ;;
let john = empathic ;;

print_endline (hug katie) ;;
print_endline (hug john) ;;
